### PR TITLE
[Proposal] Introduce OutcomeArguments

### PR DESCRIPTION
**Pull request by martintmk**
### Details on the issue fix or feature implementation

Came up in the API review.

These could simplify the delegate definition from:

``` csharp
new ResilienceStrategyBuilder<HttpResponseMessage>()
    .AddRetry(new RetryStrategyOptions<HttpResponseMessage>
    {
        ShouldRetry = (outcome, args)  => outcome switch
        {
            { Exception: HttpRequestException } => PredicateResult.True,
            { Result: HttpResponseMessage response } when !response.IsSuccessStatusCode => PredicateResult.True,
            _ => PredicateResult.False
        }
    });
```

To:

``` csharp
new ResilienceStrategyBuilder<HttpResponseMessage>()
    .AddRetry(new RetryStrategyOptions<HttpResponseMessage>
    {
        ShouldRetry = outcome => outcome switch
        {
            { Exception: HttpRequestException } => PredicateResult.True,
            { Result: HttpResponseMessage response } when !response.IsSuccessStatusCode => PredicateResult.True,
            _ => PredicateResult.False
        }
    });
```

Slightly simpler syntax at expense of one extra type. Would still allow us to pass the individual arguments to telemetry. The `IResilienceArguments` interface won't have much use anymore.

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

**Comment by martintmk** [comment]
@martincostello , @joelhulen what's your opinion here? Should I give this a shot?

File: src/Polly.Core/Strategy/OutcomeArguments.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    /// Encapsulates the outcome of a resilience operation or event and its associated arguments.
+    /// </summary>
+    /// <typeparam name="TResult">The type of the result returned by the resilience operation or event.</typeparam>
+    /// <typeparam name="TArgs">The type of the additional arguments associated with the specific resilience operation or event.</typeparam>
+    public readonly struct OutcomeArguments<TResult, TArgs>
+#pragma warning restore CA1815 // Override equals and operator equals on value types
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OutcomeArguments{TResult, TArgs}"/> struct.
+        /// Creates a new instance of <see cref="OutcomeArguments{TResult, TArgs}"/>.
```
Comment (martincostello) [inline src/Polly.Core/Strategy/OutcomeArguments.cs:15]: We don't need both sentences here?

File: src/Polly.RateLimiting.Tests/RateLimiterResilienceStrategyTests.cs [hunk 2]
```diff
@@ -47,7 +47,7 @@ public void Execute_HappyPath()
     public void Execute_LeaseRejected(bool hasEvents, bool hasRetryAfter)
     {
         _diagnosticSource.Setup(v => v.IsEnabled("OnRateLimiterRejected")).Returns(true);
-        _diagnosticSource.Setup(v => v.Write("OnRateLimiterRejected", It.Is<object>(obj => obj is IResilienceArguments)));
+        _diagnosticSource.Setup(v => v.Write("OnRateLimiterRejected", It.Is<object>(obj => obj != null)));
```
Comment (martincostello) [inline src/Polly.RateLimiting.Tests/RateLimiterResilienceStrategyTests.cs:50]:
```suggestion
        _diagnosticSource.Setup(v => v.Write("OnRateLimiterRejected", It.Is<object>(obj => obj is not null)));
```

**Review by martincostello — APPROVED** [review]

File: src/Polly.RateLimiting.Tests/RateLimiterResilienceStrategyTests.cs (hunk 2 above)
Comment (martintmk) [reply src/Polly.RateLimiting.Tests/RateLimiterResilienceStrategyTests.cs:50]: It complains that expression tree cannot contain is operator. I have to leave it as it is.

**Pull request merged**